src/App.js

```diff
@@ -2,22 +2,35 @@ import React, { useEffect, useState } from "react";
 import AddTodoForm from "./AddTodoForm";
 import TodoList from "./TodoList";
 
+
 function App() {
-  const [todoList, setTodoList] = useSemiPersistentState();
+  const [todoList, setTodoList] = useState([]); //updated to empty array
+  const [isLoading, setIsLoading] = useState(true)
+
+  useEffect(() =>{
+    const myPromise = new Promise((resolve, reject) => {
+      setTimeout(() =>{
+        resolve({data: {todoList: JSON.parse(localStorage.getItem("SavedTodoList")) ?? []}})
+      }, 2000)
+    })
+    myPromise.then((result) =>{
+        setTodoList(result.data.todoList);
+        setIsLoading(false)
+    })
+   }, [])
+
+  useEffect(() => {
+    if(!isLoading){
+        localStorage.setItem("SavedTodoList", JSON.stringify(todoList));
+    }
+    }, [todoList, isLoading]);
+    
+
   const addTodo = (newTodo) => {
     setTodoList([...todoList, newTodo]);
```

Comment:
While this works as intended, there's a bug in how its implemented. To be brief, React batches state updates. Therefore, if you have something like this:
```js
// const [count, setCount] = useState(0);
setCount(count + 1) // expected 0 + 1
setCount(count + 1) // expected 1 + 1
setCount(count + 1) // expected 2 + 1
// expected count = 3
```
You may think the count is being set to 3, but it's not! Instead, since React batches those state setters, it takes the current count (0) and increases it by 1 multiple times:
```js
// const [count, setCount] = useState(0);
setCount(count + 1) // actual 0 + 1
setCount(count + 1) // actual 0 + 1
setCount(count + 1) // actual 0 + 1
// actual count = 1
```

To mitigate this bug, you'll want to use a **function** inside of the state setter whose first argument is the previous state (`prevCount`) and returns the next state (`prevCount + 1`):
```js
// const [count, setCount] = useState(0);
setCount(prevCount => prevCount + 1) // actual 0 + 1
setCount(prevCount => prevCount + 1) // actual 1 + 1
setCount(prevCount => prevCount + 1) // actual 2 + 1
// actual count = 3
```
---
With the above knowledge, how might you change this to reflect the above?
```js
setTodoList([...todoList,newTodo])
```

<details>
<summary>Click here for answer if you're stuck</summary>

```js
setTodoList(prevTodoList => ([ ...prevTodoList, newTodo ]))
```
</details>

Reply:
Thank you!!